Custom inserts with Query annotation not working correctly

Custom inserts result in exception:

```
Caused by: io.micronaut.data.exceptions.DataAccessException: Supplied entity is a transient instance: io.micronaut.data.jdbc.h2.EnumEntity@245253d8
	at io.micronaut.data.runtime.operations.internal.AbstractSqlRepositoryOperations.setStatementParameter(AbstractSqlRepositoryOperations.java:932)
	at io.micronaut.data.runtime.operations.internal.AbstractSqlRepositoryOperations$StoredSqlOperation.setStatementParameter(AbstractSqlRepositoryOperations.java:1782)
```

### Steps to Reproduce

Remove `@PendingFeature` from https://github.com/micronaut-projects/micronaut-data/pull/996/files#diff-615d9cfd14608ba922c3fa5817b49c9a1ec7d432154f456fa90075dd718ca515R42-R46

--- a/data-hibernate-jpa/src/test/java/io/micronaut/data/hibernate/PersonCrudRepository.java
+++ b/data-hibernate-jpa/src/test/java/io/micronaut/data/hibernate/PersonCrudRepository.java
@@ -15,6 +15,7 @@
  */
 package io.micronaut.data.hibernate;
 
+import io.micronaut.context.annotation.Parameter;
 import io.micronaut.data.annotation.Query;
 import io.micronaut.data.annotation.Repository;
 import io.micronaut.data.jpa.repository.JpaRepository;
@@ -29,6 +30,13 @@ import java.util.List;
 
 @Repository
 public interface PersonCrudRepository extends JpaRepository<Person, Long>, PersonRepository {
+
+    @Override
+    Person save(@Parameter("name") String name, @Parameter("age") int age);
+
+    @Override
+    void saveCustom(@Parameter("name") String name, @Parameter("age") int age);
+
     @Query("from Person p where p.name = :n")
     @Transactional
     List<Person> listPeople(String n);

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2EnumsMappingSpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2EnumsMappingSpec.groovy
@@ -39,7 +39,6 @@ class H2EnumsMappingSpec extends Specification implements H2TestPropertyProvider
     @Inject
     JpaEnumEntityRepository jpaEnumEntityRepository = applicationContext.getBean(JpaEnumEntityRepository)
 
-    @PendingFeature(reason = "Investigate why custom inserts don't work")
     void 'test inserts are broken for custom queries'() {
         expect:
         enumEntityRepository.insertValue("b", "b", 1)
@@ -151,7 +150,7 @@ abstract class EnumEntityRepository implements CrudRepository<EnumEntity, Long> 
         })
     }
 
-    @Query("INSERT INTO ENUM_ENTITY(as_default, as_string, as_int) VALUES(:asDefault,:asString, :asString)")
+    @Query("INSERT INTO ENUM_ENTITY(as_default, as_string, as_int) VALUES(:asDefault,:asString,:asInt)")
     abstract void insertValue(String asDefault, String asString, int asInt)
 
     abstract EnumEntity findByAsString(MyEnum value)

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/H2PersonRepository.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/H2PersonRepository.java
@@ -15,9 +15,17 @@
  */
 package io.micronaut.data.jdbc.h2;
 
+import io.micronaut.data.annotation.Query;
 import io.micronaut.data.jdbc.annotation.JdbcRepository;
 import io.micronaut.data.model.query.builder.sql.Dialect;
+import io.micronaut.data.tck.entities.Person;
 
 @JdbcRepository(dialect = Dialect.H2)
 public interface H2PersonRepository extends io.micronaut.data.tck.repositories.PersonRepository {
+
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, TRUE)")
+    void saveCustom(String name, int age);
+
 }

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/mysql/MySqlPersonRepository.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/mysql/MySqlPersonRepository.java
@@ -15,9 +15,17 @@
  */
 package io.micronaut.data.jdbc.mysql;
 
+import io.micronaut.data.annotation.Query;
 import io.micronaut.data.jdbc.annotation.JdbcRepository;
 import io.micronaut.data.model.query.builder.sql.Dialect;
+import io.micronaut.data.tck.entities.Person;
 
 @JdbcRepository(dialect = Dialect.MYSQL)
 public interface MySqlPersonRepository extends io.micronaut.data.tck.repositories.PersonRepository {
+
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, TRUE)")
+    void saveCustom(String name, int age);
+
 }

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/oraclexe/OracleXEPersonRepository.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/oraclexe/OracleXEPersonRepository.java
@@ -26,6 +26,11 @@ import java.util.List;
 @JdbcRepository(dialect = Dialect.ORACLE)
 public interface OracleXEPersonRepository extends PersonRepository {
 
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(id, name, age, enabled) VALUES (\"PERSON_SEQ\".nextval, :name, :age, 1)")
+    void saveCustom(String name, int age);
+
     @Query("INSERT INTO person(id, name, age, enabled) VALUES (\"PERSON_SEQ\".nextval, :name, :age, 1)")
     void saveCustom(List<Person> people);
 

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/postgres/PostgresPersonRepository.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/postgres/PostgresPersonRepository.java
@@ -15,9 +15,17 @@
  */
 package io.micronaut.data.jdbc.postgres;
 
+import io.micronaut.data.annotation.Query;
 import io.micronaut.data.jdbc.annotation.JdbcRepository;
 import io.micronaut.data.model.query.builder.sql.Dialect;
+import io.micronaut.data.tck.entities.Person;
 
 @JdbcRepository(dialect = Dialect.POSTGRES)
 public interface PostgresPersonRepository extends io.micronaut.data.tck.repositories.PersonRepository {
+
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, TRUE)")
+    void saveCustom(String name, int age);
+
 }

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/sqlserver/MSSQLPersonRepository.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/sqlserver/MSSQLPersonRepository.java
@@ -26,6 +26,11 @@ import java.util.List;
 @JdbcRepository(dialect = Dialect.SQL_SERVER)
 public interface MSSQLPersonRepository extends PersonRepository {
 
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, 1)")
+    void saveCustom(String name, int age);
+
     @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, 1)")
     void saveCustom(List<Person> people);
 

--- a/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/RawQueryMethod.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/RawQueryMethod.java
@@ -78,10 +78,7 @@ public class RawQueryMethod implements MethodCandidate {
     @Override
     public MethodMatchInfo buildMatchInfo(@NonNull MethodMatchContext matchContext) {
         MethodElement methodElement = matchContext.getMethodElement();
-        RawQuery rawQuery = buildRawQuery(matchContext);
-        if (rawQuery == null) {
-            return null;
-        }
+
         ParameterElement[] parameters = matchContext.getParameters();
         ParameterElement entityParameter = Arrays.stream(parameters).filter(p -> TypeUtils.isEntity(p.getGenericType())).findFirst().orElse(null);
         ParameterElement entitiesParameter = Arrays.stream(parameters).filter(p -> TypeUtils.isIterableOfEntity(p.getGenericType())).findFirst().orElse(null);
@@ -89,6 +86,11 @@ public class RawQueryMethod implements MethodCandidate {
         boolean readOnly = matchContext.getAnnotationMetadata().booleanValue(Query.class, "readOnly").orElse(true);
         String query = matchContext.getAnnotationMetadata().stringValue(Query.class).get();
         MethodMatchInfo.OperationType operationType = findOperationType(methodElement.getName(), query, readOnly);
+
+        RawQuery rawQuery = buildRawQuery(matchContext, operationType);
+        if (rawQuery == null) {
+            return null;
+        }
 
         Map.Entry<ClassElement, Class<? extends DataInterceptor>> entry = FindersUtils.resolveInterceptorTypeByOperationType(
                 entityParameter != null,
@@ -114,7 +116,7 @@ public class RawQueryMethod implements MethodCandidate {
             }
         }
 
-        QueryModel queryModel = buildRawQuery(matchContext);
+        QueryModel queryModel = buildRawQuery(matchContext, operationType);
         MethodMatchInfo methodMatchInfo;
         if (isDto) {
             methodMatchInfo = new MethodMatchInfo(
@@ -169,9 +171,10 @@ public class RawQueryMethod implements MethodCandidate {
      * Builds a raw query for the given match context. Should be called for methods annotated with {@link Query} explicitly.
      *
      * @param matchContext The match context
+     * @param operationType The operation type
      * @return The raw query or null if an error occurred
      */
-    private RawQuery buildRawQuery(@NonNull MethodMatchContext matchContext) {
+    private RawQuery buildRawQuery(@NonNull MethodMatchContext matchContext, MethodMatchInfo.OperationType operationType) {
         MethodElement methodElement = matchContext.getMethodElement();
         String queryString = methodElement.stringValue(Query.class).orElseThrow(() ->
                 new IllegalStateException("Should only be called if Query has value!")
@@ -229,7 +232,8 @@ public class RawQueryMethod implements MethodCandidate {
                 }
             }
         }
-        return new RawQuery(matchContext.getRootEntity(), parameterBindings, persistentEntity != null);
+        boolean encodeEntityParameters = persistentEntity != null || operationType == MethodMatchInfo.OperationType.INSERT;
+        return new RawQuery(matchContext.getRootEntity(), parameterBindings, encodeEntityParameters);
     }
 
 }

--- a/data-r2dbc/src/test/groovy/io/micronaut/data/r2dbc/h2/H2TransactionManagementSpec.groovy
+++ b/data-r2dbc/src/test/groovy/io/micronaut/data/r2dbc/h2/H2TransactionManagementSpec.groovy
@@ -14,7 +14,14 @@ import javax.inject.Inject
 class H2TransactionManagementSpec extends Specification implements H2TestPropertyProvider {
 
     @Shared @Inject H2OwnerRepository ownerRepository
-    @Shared @Inject H2PetRepository petRepository
+
+    def setup() {
+        Mono.from(ownerRepository.deleteAll()).block()
+    }
+
+    def cleanup() {
+        Mono.from(ownerRepository.deleteAll()).block()
+    }
 
     void 'test rollback only'() {
         when:"When setRollbackOnly is called"

--- a/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/h2/H2PersonRepository.java
+++ b/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/h2/H2PersonRepository.java
@@ -18,10 +18,17 @@ package io.micronaut.data.r2dbc.h2;
 import io.micronaut.data.annotation.Query;
 import io.micronaut.data.model.query.builder.sql.Dialect;
 import io.micronaut.data.r2dbc.annotation.R2dbcRepository;
+import io.micronaut.data.tck.entities.Person;
 import io.micronaut.data.tck.entities.TotalDto;
 
 @R2dbcRepository(dialect = Dialect.H2)
 public interface H2PersonRepository extends io.micronaut.data.tck.repositories.PersonRepository {
+
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, TRUE)")
+    void saveCustom(String name, int age);
+
     @Query("select count(*) as total from person")
     TotalDto getTotal();
 }

--- a/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/mysql/MySqlPersonRepository.java
+++ b/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/mysql/MySqlPersonRepository.java
@@ -15,9 +15,17 @@
  */
 package io.micronaut.data.r2dbc.mysql;
 
+import io.micronaut.data.annotation.Query;
 import io.micronaut.data.model.query.builder.sql.Dialect;
 import io.micronaut.data.r2dbc.annotation.R2dbcRepository;
+import io.micronaut.data.tck.entities.Person;
 
 @R2dbcRepository(dialect = Dialect.MYSQL)
 public interface MySqlPersonRepository extends io.micronaut.data.tck.repositories.PersonRepository {
+
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, TRUE)")
+    void saveCustom(String name, int age);
+
 }

--- a/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/oraclexe/OracleXEPersonRepository.java
+++ b/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/oraclexe/OracleXEPersonRepository.java
@@ -26,6 +26,11 @@ import java.util.List;
 @R2dbcRepository(dialect = Dialect.ORACLE)
 public interface OracleXEPersonRepository extends PersonRepository {
 
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(id, name, age, enabled) VALUES (\"PERSON_SEQ\".nextval, :name, :age, 1)")
+    void saveCustom(String name, int age);
+
     @Query("INSERT INTO person(id, name, age, enabled) VALUES (\"PERSON_SEQ\".nextval, :name, :age, 1)")
     void saveCustom(List<Person> people);
 

--- a/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/postgres/PostgresPersonRepository.java
+++ b/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/postgres/PostgresPersonRepository.java
@@ -15,10 +15,17 @@
  */
 package io.micronaut.data.r2dbc.postgres;
 
+import io.micronaut.data.annotation.Query;
 import io.micronaut.data.model.query.builder.sql.Dialect;
 import io.micronaut.data.r2dbc.annotation.R2dbcRepository;
+import io.micronaut.data.tck.entities.Person;
 
 @R2dbcRepository(dialect = Dialect.POSTGRES)
 public interface PostgresPersonRepository extends io.micronaut.data.tck.repositories.PersonRepository {
+
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, TRUE)")
+    void saveCustom(String name, int age);
     
 }

--- a/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/sqlserver/MSPersonRepository.java
+++ b/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/sqlserver/MSPersonRepository.java
@@ -26,6 +26,11 @@ import java.util.List;
 @R2dbcRepository(dialect = Dialect.SQL_SERVER)
 public interface MSPersonRepository extends PersonRepository {
 
+    Person save(String name, int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, 1)")
+    void saveCustom(String name, int age);
+
     @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, 1)")
     void saveCustom(List<Person> people);
 

--- a/data-tck/src/main/groovy/io/micronaut/data/tck/tests/AbstractRepositorySpec.groovy
+++ b/data-tck/src/main/groovy/io/micronaut/data/tck/tests/AbstractRepositorySpec.groovy
@@ -360,6 +360,24 @@ abstract class AbstractRepositorySpec extends Specification {
         isOracle() || personRepository.list(Pageable.from(0, 1)).size() == 1
     }
 
+    void "test save many custom"() {
+        given:
+        savePersons(["Jeff", "James"])
+
+        when:"many are saved"
+        personRepository.saveCustom("Frank", 0)
+        personRepository.saveCustom("Bob", 0)
+
+        then:"all are saved"
+        personRepository.findAll().size() == 4
+        personRepository.count() == 4
+        personRepository.count("Jeff") == 1
+
+        // Oracle 11g doesn't support pagination
+        isOracle() || personRepository.list(Pageable.from(1)).isEmpty()
+        isOracle() || personRepository.list(Pageable.from(0, 1)).size() == 1
+    }
+
     void "test update many"() {
         given:
         savePersons(["Jeff", "James"])

--- a/data-tck/src/main/java/io/micronaut/data/tck/repositories/PersonRepository.java
+++ b/data-tck/src/main/java/io/micronaut/data/tck/repositories/PersonRepository.java
@@ -43,7 +43,10 @@ public interface PersonRepository extends CrudRepository<Person, Long>, Pageable
 
     int countByAgeLessThan(int wrapper);
 
-    Person save(String name, int age);
+    Person save(@Parameter("name") String name, @Parameter("age") int age);
+
+    @Query("INSERT INTO person(name, age, enabled) VALUES (:name, :age, TRUE)")
+    void saveCustom(@Parameter("name") String name, @Parameter("age") int age);
 
     Person get(Long id);
 

--- a/doc-examples/r2dbc-example-groovy/src/test/groovy/example/BookControllerTest.groovy
+++ b/doc-examples/r2dbc-example-groovy/src/test/groovy/example/BookControllerTest.groovy
@@ -8,7 +8,6 @@ import io.micronaut.test.extensions.spock.annotation.MicronautTest
 import io.micronaut.test.support.TestPropertyProvider
 import io.reactivex.Flowable
 import org.testcontainers.containers.MySQLContainer
-import org.testcontainers.utility.DockerImageName
 import reactor.core.publisher.Flux
 import reactor.core.publisher.Mono
 import spock.lang.Shared
@@ -68,7 +67,7 @@ class BookControllerTest extends Specification implements TestPropertyProvider {
 
     @Override
     Map<String, String> getProperties() {
-        container = new MySQLContainer<>(DockerImageName.parse("mysql").withTag("5"))
+        container = new MySQLContainer<>("mysql:8.0.17")
         container.start()
         return CollectionUtils.mapOf(
                 "datasources.default.url", container.getJdbcUrl(),

--- a/doc-examples/r2dbc-example-java/src/test/java/example/BookControllerTest.java
+++ b/doc-examples/r2dbc-example-java/src/test/java/example/BookControllerTest.java
@@ -10,7 +10,6 @@ import io.reactivex.Flowable;
 import io.reactivex.Single;
 import org.junit.jupiter.api.*;
 import org.testcontainers.containers.MySQLContainer;
-import org.testcontainers.utility.DockerImageName;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
@@ -92,7 +91,7 @@ public class BookControllerTest implements TestPropertyProvider {
 
     @Override
     public Map<String, String> getProperties() {
-        container = new MySQLContainer<>(DockerImageName.parse("mysql").withTag("5"));
+        container = new MySQLContainer<>("mysql:8.0.17");
         container.start();
         return CollectionUtils.mapOf(
                 "datasources.default.url", container.getJdbcUrl(),

--- a/doc-examples/r2dbc-example-kotlin/src/test/java/example/MySqlServer.java
+++ b/doc-examples/r2dbc-example-kotlin/src/test/java/example/MySqlServer.java
@@ -1,12 +1,11 @@
 package example;
 
 import org.testcontainers.containers.MySQLContainer;
-import org.testcontainers.utility.DockerImageName;
 
 public class MySqlServer {
 
     static MySQLContainer<?> start() {
-        MySQLContainer<?> container = new MySQLContainer<>(DockerImageName.parse("mysql").withTag("5"));
+        MySQLContainer<?> container = new MySQLContainer<>("mysql:8.0.17");
         container.start();
         return container;
     }

